setup/helpers/abs_helper.rb

```diff
@@ -0,0 +1,284 @@
+          puts "#{Time.now} - Retrying ABS request..."
+          sleep(time_increment)
+        else
+          break
+        end
+      end
+    end
+    res
+  end
+
+  def abs_get_request_post(uri)
+    Net::HTTP::Post.new(uri, {'Content-Type' => 'application/json', 'X-Auth-Token' => abs_get_token})
+  end
+
+  def abs_request_awsdirect(uri, body)
+    req = abs_get_request_post(uri)
+    req.body = body
+    http = Net::HTTP.new(uri.host, uri.port)
+    http.use_ssl = true
+    http.read_timeout = 120
+
+    puts 'sending request body:'
+    puts body
+    puts "to uri: #{uri}"
+    puts
+
+    res = http.request(req)
+
+    puts "response: #{res.body}" unless res.nil?
+
+    res
+  end
+
+  def abs_is_valid_response?(res)
+    # TODO: other criteria?
+    invalid_response_bodies = ['', nil]
+    valid_response_codes = ['200']
+    is_valid_response = false
+
+    if res.nil? || !valid_response_codes.include?(res.code) || invalid_response_bodies.include?(res.body)
+      puts 'Invalid ABS response: '
+      puts "code: #{res.code}" unless res.nil?
+      puts "body: #{res.body}" unless res.nil?
+    else
+      is_valid_response = true
+    end
+    is_valid_response
+  end
+
+  def abs_get_a2a_hosts
+    # TODO: A2A_HOSTS environment variable?
+    {'mom': abs_get_aws_mom_size, 'metrics': abs_get_aws_metrics_size}
+  end
+
+  def abs_reformat_resource_host(response_body)
+    reformatted_json = nil
+
+    begin
+      json = JSON.parse(response_body)
+      hostname = json['hostname']
+      type = json['type']
+
+      new_response_body = {
+          'hostname': hostname,
+          'type':     ABS_BEAKER_TYPE,
+          'engine':   ABS_BEAKER_ENGINE}
+
+      reformatted_json = new_response_body.to_json
+    rescue
+      puts 'JSON::ParserError encountered'
+    end
+
+    reformatted_json
+  end
+
+  def abs_get_resource_hosts
+    # Allows us to switch between AWS and VMPooler by selecting different ABS os's
+    # centos-7-x86-64-west is an AWS image, centos-7-x86_64 is vmpooler
+    uri = URI("#{ABS_BASE_URL}/awsdirect")
+    responses = []
+    abs_resource_hosts = nil
+    invalid_response = false
+
+    hosts = abs_get_a2a_hosts
+
+    hosts.each do |role, size|
+
+      request_body = abs_get_awsdirect_request_body(role, size)
+      response = abs_request_awsdirect(uri, request_body)
+
+      # if any invalid responses are encountered stop and return any provisioned hosts
+      if !abs_is_valid_response?(response)
+        invalid_response = true
+        puts "Unable to provision host for role: #{role}"
+
+        puts 'Returning any provisioned hosts'
+        if !responses.empty?
+          ENV['ABS_RESOURCE_HOSTS'] = responses.to_json
+          abs_return_resource_hosts
+        end
+
+        # stop requesting hosts
+        break
+
+      else
+        # reformat to satisfy beaker
+        responses << JSON.parse(abs_reformat_resource_host(response.body))
+      end
+
+    end
+
+    if responses.empty?
```

Comment:
This feels odd.  The only way to get here I think is if the first request returned nil.  Which would be an invalid response.  Feels like this should be if invalid response then print message about any a provisioning error and that any provisions hosts were terminated.  But That messaging is already there I think.  Unless nil isn't considered an invalid response.  So I think this if part can be tossed, and the elseif converted to an if.

Reply:
Actually, not sure my comment addressed your point. Looking into it...

Reply:
Looking at this again, the responses array could be empty if either no requests were made (no hosts specified or no token) or if the first response was invalid. If the responses array is empty we don't need to proceed further since no hosts were provisioned. And we don't want to check whether invalid_response is true independently; it may actually be false if no request was made. 

Should this be refactored?

Reply:
I've updated this to check for a token via abs_initialize before proceeding and moved the empty responses message to where it seems to make more sense.